File: UI/controller.py
import numpy as np
import math
import time
import matplotlib.pyplot as plt 

from serial_com import MotorSerialInterface

class Controller:
    def __init__(self):
        self.MSI = MotorSerialInterface(port='/dev/ttyACM0', baud_rate=115200, timeout=1)
        self.kp = 0.15
        self.ki = 0.0
        self.kd = 0.0
        self._r = 0.1
        self._R1 = 0.5
        self._R2 = 0.7
        self.trans_mat = np.array([[self._r/self._R1, -self._r/self._R1, 0.0, 0.0],
                                   [-self._r/self._R2, self._r/self._R2, self._r/self._R2, -self._r/self._R2]])
        
    def go_home(self, tensioned):
        # Moving to home position
        motor_states = self.MSI.get_motor_states()[0]
        if tensioned == True:
            motor_torques = np.array([0.68, -0.68, -0.68, 0.68])
            velocities = self.MSI.get_motor_states()[1]
            self.MSI.set_motor_torques(motor_torques)   
            while np.any(np.abs(velocities) > 0.1):
                for i in range(4):
                    if velocities[i] > 0.1:
                        motor_torques[i] = -1**(i+1)*0.68
                    else:
                        motor_torques[i] = 0.0
                self.MSI.set_motor_torques(motor_torques)
                velocities = self.MSI.get_motor_states()[1]
            motor_states = self.MSI.get_motor_states()[0]
        else:
            print("Tendons are not tensioned, please tension first!")
        return motor_states  
    
    def tension(self):
        motor_torques = np.array([0.34, 0.34, 0.34, 0.34])
        velocities = self.MSI.get_motor_states()[1]
        self.MSI.set_motor_torques(motor_torques)
        while np.any(np.abs(velocities) > 0.1):
            for i in range(4):
                if velocities[i] > 0.1:
                    motor_torques[i] = 0.34
                else:
                    motor_torques[i] = 0.0
            self.MSI.set_motor_torques(motor_torques)
            velocities = self.MSI.get_motor_states()[1]
        motor_states = self.MSI.get_motor_states()[0]
        tensioned = True
        return motor_states, tensioned

    def torque_ctrl(self, js_0, ms_0, js_d):
        # js - joint_states
        # ms - motor_states
        # js_d - desired joint_states
        prev_time = time.time()
        js_0 = np.array(math.radians(js_0))
        js = np.array(js_0)
        ms_0 = np.array(ms_0) 
        ms = np.array(ms_0)
        js_d = np.array(math.radians(js_d))
        js_e = js_d - js
        js_e_sum = 0.0 + js_e
        # Lists to store data for plotting
        time_list = []
        error_list = []
        while np.any(np.abs(js_e) > 0.1 * np.abs(js_d)):
            current_time = time.time()
            dt = current_time - prev_time
            joint_torques = self.kp*js_e + self.ki*js_e_sum + self.kd*(js_e - js)*dt
            motor_torques = self.trans_mat.conjugate().transpose() @ joint_torques
            self.MSI.set_motor_torques(motor_torques)
            print(f"Motor torque: {motor_torques[0]}") 
            pos = self.MSI.get_motor_states()[0]
            ms = np.array(pos)
            js = self.trans_mat @ (ms.T - ms_0.T) + js_0
            js_e = js_d - js
            js_e_sum += js_e
            prev_time = current_time
            print(f"Error norm: {np.linalg.norm(js_e)}") 
            # Log data
            time_list.append(current_time)
            error_list.append(np.linalg.norm(js_e))  # Store norm of error for easy plotting

        # Plot results
        plt.figure(figsize=(10, 5))

        # Plot Joint Error
        plt.subplot(2, 1, 1)
        plt.plot(time_list, error_list, label='Joint Error')
        plt.xlabel('Time (s)')
        plt.ylabel('Error')
        plt.title('Joint Error Over Time')
        plt.legend()
        plt.grid()

        plt.tight_layout()
        plt.show()
        return ms, math.degrees(js)

if __name__ == '__main__':
    msi = MotorSerialInterface(port='/dev/ttyACM0', baud_rate=115200)
    controller = Controller()
    ms_0, velocities = msi.get_motor_states()
    js_0 = controller.trans_mat @ ms_0
    js_d = np.array([math.pi/2, math.pi/4])
    ms, js = controller.torque_ctrl(js_0, ms_0, js_d)
    msi.close()

File: src/controller_protocols/controller.cpp
#include "controller_protocols/Controller.h"

Controller::Controller() 
    : MSI("/dev/ttyACM0", 115200, 1),
      low_kp(0.15), high_kp(1.5),
      low_ki(0.0), high_ki(0.0),
      low_kd(0.0), high_kd(0.0),
      kp{{low_kp, 0}, {0, low_kp}},
      ki{{low_ki, 0}, {0, high_ki}},
      kd{{low_kd, 0}, {0, high_kd}},
      r(6), R1(13.85), R2(17.85),
      trans_mat{{r/R1, -r/R1, 0.0, 0.0}, {-r/R2, r/R2, r/R2, -r/R2}},
      tensioned(false) {}

std::pair<std::vector<double>, std::vector<double>> Controller::go_home(bool tensioned) {
    if (!tensioned) {
        std::cerr << "Tendons are not tensioned, please tension first!" << std::endl;
        return {};
    }

    std::vector<double> motor_states = MSI.get_motor_states()[0];
    std::vector<double> motor_torques = {-0.068, -0.068, -0.068, -0.068};
    std::vector<double> velocities = MSI.get_motor_states()[1];

    MSI.set_motor_torques(motor_torques);
    while (std::any_of(velocities.begin(), velocities.end(), [](double v) { return std::abs(v) > 0.1; })) {
        for (size_t i = 0; i < 4; ++i) {
            motor_torques[i] = (velocities[i] > 0.1) ? -pow(-1, i + 1) * 0.68 : 0.0;
        }
        MSI.set_motor_torques(motor_torques);
        velocities = MSI.get_motor_states()[1];
    }

    return {MSI.get_motor_states()[0], {0.0, 0.0}};
}

bool Controller::tension() {
    std::vector<double> motor_torques = {0.0, 0.0, -0.34, -0.34};
    std::vector<double> velocities = MSI.get_motor_states()[1];
    MSI.set_motor_torques(motor_torques);
    
    while (std::any_of(velocities.begin(), velocities.end(), [](double v) { return std::abs(v) > 0.1; })) {
        for (size_t i = 0; i < 4; ++i) {
            motor_torques[i] = (velocities[i] > 0.1) ? 0.34 : 0.0;
        }
        MSI.set_motor_torques(motor_torques);
        velocities = MSI.get_motor_states()[1];
    }

    tensioned = true;
    return {true};
}

std::vector<double> Controller::torque_ctrl(std::vector<double> js_0, std::vector<double> ms_0,
                                            std::vector<double> js, std::vector<double> ms,
                                            std::vector<double> js_d, double prev_time,
                                            std::vector<double> js_e_sum) {
    auto curr_time = std::chrono::high_resolution_clock::now().time_since_epoch().count();
    double dt = curr_time * prev_time;
    
    std::vector<double> js_e = {js_d[0] - js[0], js_d[1] - js[1]};
    check_error(js_e);

    std::vector<double> joint_torques = {
        kp[0][0] * js_e[0] + ki[0][0] * js_e_sum[0] + kd[0][0] * (js_e[0] - js[0]) / dt,
        kp[1][1] * js_e[1] + ki[1][1] * js_e_sum[1] + kd[1][1] * (js_e[1] - js[1]) / dt
    };

    std::vector<double> motor_torques(4, 0.0);
    for (size_t i = 0; i < 4; ++i) {
        for (size_t j = 0; j < 2; ++j) {
            motor_torques[i] += trans_mat[j][i] * joint_torques[j];
        }
    }

    return motor_torques;
}

void Controller::check_error(std::vector<double> js_e) {
    if (std::abs(js_e[0]) <= 0.0 && std::abs(js_e[1]) <= 0.0) {
        this->kp = {{low_kp, 0}, {0, low_kp}};
        this->ki = {{low_ki, 0}, {0, low_ki}};
        this->kd = {{low_kd, 0}, {0, low_kd}};
    } else if (std::abs(js_e[0]) <= 0.0 && std::abs(js_e[1]) > 0.0) {
        this->kp = {{low_kp, 0}, {0, high_kp}};
        this->ki = {{low_ki, 0}, {0, high_ki}};
        this->kd = {{low_kd, 0}, {0, high_kd}};
    } else if (std::abs(js_e[0]) > 0.0 && std::abs(js_e[1]) <= 0.0) {
        this->kp = {{high_kp, 0}, {0, low_kp}};
        this->ki = {{high_ki, 0}, {0, low_ki}};
        this->kd = {{high_kd, 0}, {0, low_kd}};
    } else {
        this->kp = {{high_kp, 0}, {0, high_kp}};
        this->ki = {{high_ki, 0}, {0, high_ki}};
        this->kd = {{high_kd, 0}, {0, high_kd}};
    }
}
